primitives/src/transaction.rs

```diff
@@ -349,6 +351,388 @@ impl Encodable for Transaction {
+                    }
+                    let version = *version;
+                    let iteration = iteration.0;
+
+                    let (mut inputs, outputs, decoder) = self.state.witness_transition();
+                    inputs[iteration].witness = decoder.end()?;
+                    if iteration < inputs.len() - 1 {
+                        self.state = State::Witnesses(
+                            version,
+                            inputs,
+                            outputs,
+                            Iteration(iteration + 1),
+                            WitnessDecoder::new(),
+                        );
+                    } else {
+                        if !inputs.is_empty() && inputs.iter().all(|input| input.witness.is_empty())
+                        {
+                            return Err(E(Inner::NoWitnesses));
+                        }
+                        self.state =
+                            State::LockTime(version, inputs, outputs, LockTimeDecoder::new());
+                    }
+                }
+                State::LockTime(version, _, _, decoder) => {
+                    if decoder.push_bytes(bytes)? {
+                        return Ok(true);
+                    }
+                    let version = *version;
+
+                    let (inputs, outputs, decoder) = self.state.lock_time_transition();
+                    let lock_time = decoder.end()?;
+                    self.state = State::Done(Transaction { version, lock_time, inputs, outputs });
+                    return Ok(false);
+                }
+                State::Done(..) => return Ok(false),
+                State::Transitioning => {
+                    panic!("use of decoder in transitioning state");
+                }
+            }
+        }
+    }
+
+    #[inline]
+    fn end(self) -> Result<Self::Output, Self::Error> {
+        use TransactionDecoderState as State;
+
+        match self.state {
+            State::Version(_) => panic!("tried to end decoder in state: Version"),
+            State::Inputs(..) => panic!("tried to end decoder in state: Inputs"),
+            State::SegwitFlag(..) => panic!("tried to end decoder in state: SegwitFlag"),
+            State::Outputs(..) => panic!("tried to end decoder in state: Outputs"),
+            State::Witnesses(..) => panic!("tried to end decoder in state: Witnesses"),
+            State::LockTime(..) => panic!("tried to end decoder in state: LockTime"),
+            State::Done(tx) => Ok(tx),
+            State::Transitioning => {
+                panic!("use of decoder in transitioning state");
+            }
+        }
+    }
+
+    #[inline]
+    fn min_bytes_needed(&self) -> usize {
+        use TransactionDecoderState as State;
+
+        match &self.state {
+            State::Version(decoder) => decoder.min_bytes_needed(),
+            State::Inputs(_, _, decoder) => decoder.min_bytes_needed(),
+            State::SegwitFlag(_) => 2,
```

Comment:
In 15b4f4c43bc790e2e86797f5dee56f7af588ba54, given how the state machine flows, I believe this should only be 1 since the marker has already been consumed and we are only analyzing the flag in this stae.

Reply:
BOOM! Good reviewing bro, you are a useful bloke to have on the team, for real.